// Copyright Â© 2020, SAS Institute Inc., Cary, NC, USA.  All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

service EnvControl {
  rpc Start(Empty) returns (ServerInfo) {};
  rpc Close(ServerInfo) returns (Empty) {};
}

message ServerInfo{
  string ip = 1;
  int32 port = 2;
}

service Env {
  rpc Handshake(Empty) returns (MetaData) {};
  rpc Make (Name) returns (Info) {};
  rpc Reset (Empty) returns (Observation) {};
  rpc Step (Action) returns (Transition) {};
  rpc Render (RenderMode) returns (RenderOut) {};
  rpc Seed (EnvSeed) returns (Empty) {};
  rpc Sample (Empty) returns (Action) {};
  rpc Close (Empty) returns (Empty) {};
}

message MetaData {
  string EnvVersion = 1;
}

message Name {
  string data = 1;
}

message Info {
  repeated Space observation_space = 1;
  repeated Space action_space = 2;
  int32 max_episode_steps = 3;
}

message Space {
  string type = 1;
  repeated int32 shape = 2;
  string dtype = 3;
  string key = 4;
  repeated float low_f = 5;
  repeated float high_f = 6;
  repeated int32 low_i = 7;
  repeated int32 high_i = 8;
  int32 n = 9;
}


message Empty {
}

message Observation {
  repeated float data_f = 1;
  repeated int32 data_i = 2;
  repeated int32 shape = 3;
}

message Action {
  repeated int32 data_f = 1;
  repeated int32 data_i = 2;
  repeated int32 shape = 3;
}

message Transition {
  Observation next_observation = 1;
  float reward = 2;
  bool done = 3;
}

message RenderMode{
    string data = 1;
}

message RenderOut {
    repeated int32 rgb = 1;
    string ansi = 2;
}

message EnvSeed{
    int32 data = 1;
}
